experimental/iterators/README.md

```diff
@@ -0,0 +1,266 @@
+`__next__()` on that `iterator` until a `StopIteration` exceptions indicates the
+end of the sequence, binding each resulting value to the loop variable. This
+roughly corresponds to:
+
+```python3
+iterable = CountToThree()
+iterator = iterable.__iter__()
+print(iterator.__next__()) # Outputs: 1
+print(iterator.__next__()) # Outputs: 2
+print(iterator.__next__()) # Outputs: 3
+print(iterator.__next__()) # Raises: StopIteration
+```
+
+#### Composability
+
+One property that makes iterators in Python so powerful is that they are
+composable. Many
+[built-in functions](https://docs.python.org/3/library/functions.html) (such as
+`enumerate`, `filter`, `map`, etc.) and standard library functions (most notably
+those defined in
+[`itertools`](https://docs.python.org/3/library/itertools.html)) accept
+arguments of type `iterable` and produce again results that are `iterable`.
+I/O is also often done with `iterable` objects (for example, via
+[`IOBase`](https://docs.python.org/3/library/io.html#io.IOBase)).
+
+Consider the following example, which computes the maximum length of a line with
+an even number of characters of all lines in a given text file:
+
+```python3
+max(
+  filter(lambda x: x % 2 == 0,
+         map(lambda x: len(x),
+             open('file.txt'))))
+```
+
+Each of the functions `max`, `filter`, `map`, and `open` either accepts or
+produces an `iterator` object (or both). The result of the three inner-most
+function is an `iterator` object that has two more `iterator` objects
+recursively nested insise of them. That object is given to `max`, which drives
+the computation by calling `__next__()` repeatably on that object, which, in
+turn, causes a cascade of calls to `__next__()` of the nested `iterator`
+objects. Since the state in each `iterator` object and the number of values
+passed between them is constant, the above example essentially works on
+arbitrarily large files.
+
+#### Distinction to Related Concepts in Python
+
+A few related but distinct concepts deserve a short discussion:
+
+* `iterable`: As mentioned before, `iterable`s in Python are things than can be
+  iterated over (such as containers like `list`s, `tuple`s, etc) while
+  `iterator`s are objects that help with that iteration. `iterable`s provide
+  `iterator`s via their `__iter__()` function. The `for` loop uses that function
+  to get an `iterator`, and uses that in turn to iterate. Confusion may arise
+  because `iterator`s *are* also `iterable` (in order to make `iterator`s usable
+  in `for` loops).
+* `generator`: This term is actually ambiguous; it may refer to
+  [`generator function`](https://docs.python.org/3/glossary.html#term-generator)
+  (which seems to be the more common usage of simply `generator`) or
+  [`generator iterator`](https://docs.python.org/3/glossary.html#term-generator-iterator).
+  In short, the former is a function that returns the latter.
+  In more detail, an `iterator function` is a function with a `yield` statement,
```

Comment:
`iterator function` just appeared out of nowhere, did you mean `generator function` (otherwise please define the new term)

Reply:
Yep, a typo, it should be `generator function` -- good catch!